From e97719d12ed9140dd33b9100596870ab9932b711 Mon Sep 17 00:00:00 2001
From: Marek Kasik <mkasik@redhat.com>
Date: Wed, 3 Jul 2024 14:44:42 +0200
Subject: [PATCH] frdp-session: Clear the update_id when the source is removed

Clear the update_id when the source is removed via return value
of its GSourceFunc so we don't try to remove it later.
---
 src/frdp-session.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/frdp-session.c b/src/frdp-session.c
index f242e1c..634f546 100644
--- a/src/frdp-session.c
+++ b/src/frdp-session.c
@@ -718,14 +718,17 @@ update (gpointer user_data)
                                            handles, ARRAYSIZE(handles));
   if (usedHandles == 0) {
       g_warning ("Failed to get FreeRDP event handle");
+      priv->update_id = 0;
       return FALSE;
   }
 
   status = WaitForMultipleObjects (usedHandles, handles, FALSE, SELECT_TIMEOUT);
   if (status == WAIT_TIMEOUT)
     return TRUE;
-  if (status == WAIT_FAILED)
+  if (status == WAIT_FAILED) {
+    priv->update_id = 0;
     return FALSE;
+  }
 
   if (!freerdp_check_event_handles (priv->freerdp_session->context)) {
     if (freerdp_get_last_error(priv->freerdp_session->context) == FREERDP_ERROR_SUCCESS) {
@@ -737,6 +740,7 @@ update (gpointer user_data)
 
   if (freerdp_shall_disconnect (priv->freerdp_session)) {
       g_idle_add ((GSourceFunc) idle_close, self);
+      priv->update_id = 0;
 
       return FALSE;
   }
-- 
2.47.1

